### PR TITLE
Compression-oriented approach to cut down time spent developing agents

### PR DESCRIPTION
**Pull request by adammitchelldev**
## What is the goal of this PR?

Significant "fat" is introduced in the implementation of agents and a lot of it follows the same if not identical patterns. In order to improve developer productivity and hopefully also make the code easier to follow, we want to follow a compression-oriented approach to the agent framework.

## What are the changes implemented in this PR?

- Refactor to using a base agent class that is instantiated per item that the agent is parallelised over.
- Automatic transaction handling.
- Refactor existing agents to use the new format.
- Refactor the `Simulation` to use the new `AgentRunner`
- Refactor agent base and common classes out of the main agents package.

### CODE REVIEW

**Comment by jmsfltchr** [comment]
This definitely helps to strip out the boilerplate and passing a lot of variables around unnecessarily

File: AgentRunnerList.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package grakn.simulation;
+
+import grakn.simulation.agents.*;
+import grakn.simulation.agents.base.AgentRunner;
+import grakn.simulation.agents.common.CityAgentRunner;
+
+public class AgentRunnerList {
+    static AgentRunner[] AGENTS = {
+            new CityAgentRunner(MarriageAgent.class),
+            new CityAgentRunner(MarriageAgent.class),
```
[the remark this answers is not in the document]
Comment (adammitchelldev) [reply AgentRunnerList.java]: Oops

Good catch

File: agents/base/Agent.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import grakn.client.GraknClient;
+import grakn.simulation.agents.World;
+
+import java.time.LocalDateTime;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+
+/**
+ * An agent that will execute some logic across the
```
Comment (haikalpribadi) [inline agents/base/Agent.java]: The sentence in this comment is not completed.

File: agents/base/AgentItem.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package grakn.simulation.agents.base;
+
+public interface AgentItem {
+    String getTracker();
+    String getSessionKey();
+}
```
Comment (haikalpribadi) [inline agents/base/AgentItem.java]: This interface feels like a very "weak" abstraction. Why is it needed? What domain concept does it really represent? Do these 2 methods need to be enforced at all times? Could there also be a better name than `*Item`?

Comment (haikalpribadi) [reply agents/base/AgentItem.java]: Is it possible that this abstraction represents a `Resource` ?

**Comment by haikalpribadi** [comment]
Overall the PR looks very good! This is not a one-off "compression-based" technique, though. ;) This is a general (and very good) practice of constantly refactoring your codebase the moment your code starts revealing abstractions that would allow you to modularise your code. So you should expect to refactor your code in this way every month or so. @adammitchelldev @jmsfltchr

**Pull request merged**